aten/src/ATen/native/cuda/ReplicationPadding.cu

```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
           gradInput.size(1)});
-        gradOutput_ = gradOutput.reshape({1, gradOutput.size(0),
+      gradOutput_ = gradOutput.reshape({1, gradOutput.size(0),
           gradOutput.size(1)});
       }
       auto devGradInput = gradInput_.packed_accessor<scalar_t, 3>();
       auto devGradOutput = gradOutput_.packed_accessor<scalar_t, 3>();
 
       int outputPlaneSize = devGradOutput.size(2);
       dim3 gridSize(THCCeilDiv(outputPlaneSize, 256),
-        devGradOutput.size(1),
-        devGradOutput.size(0));
+          devGradOutput.size(1),
+          devGradOutput.size(0));
       dim3 blockSize(outputPlaneSize > 256 ? 256 : outputPlaneSize);
 
       replication_pad_backward_kernel <<<gridSize, blockSize, 0,
-        at::cuda::getCurrentCUDAStream()>>>(devGradInput, devGradOutput,
-            padL, padR);
+                                      at::cuda::getCurrentCUDAStream()>>>(devGradInput, devGradOutput,
+                                          padL, padR);
+      }
+  );
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void replication_pad2d_out_cuda_template(
+    Tensor& output,
+    const Tensor& input,
+    IntList paddingSize)
+{
+  AT_CHECK(at::cuda::detail::canUse32BitIndexMath(input),
+      "input tensor must fit into 32-bit index math");
+  AT_CHECK(paddingSize.size() == 4, "padding Size is expected to be 4");
+
+  int padL = paddingSize[0];
+  int padR = paddingSize[1];
+  int padT = paddingSize[2];
+  int padB = paddingSize[3];
+  int planeDim = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int numBatch = 1;
+
+  int numInputDims = input.dim();
+  AT_CHECK(input.numel() && (numInputDims == 3 || numInputDims == 4),
+      "non-empty 3D or 4D (batch mode) tensor expected for input, but got: ",
+      input)
+
+    if (numInputDims == 4) {
+      numBatch = input.size(0);
+      planeDim++;
+      dimh++;
+      dimw++;
     }
+
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(dimh);
+  int inputW = input.size(dimw);
+  int outputH = inputH + padT + padB;
+  int outputW  = inputW + padL + padR;
+
+  AT_CHECK(outputW >= 1 || outputH >= 1,
+      "input (H: ", inputH, ", W: ", inputW, ") is too small."
+      " Calculated output H: ", outputH, " W: ", outputW);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      input.type(), "replication_pad2d", [&] {
+
+
+      if (numInputDims == 3) {
+        output.resize_({numPlanes, outputH, outputW});
+        auto input_ = input.reshape({1, input.size(0), input.size(1), input.size(2)});
+        auto output_ = output.reshape({1, output.size(0), output.size(1), output.size(2)});
+        auto devInput = input_.packed_accessor<scalar_t, 4>();
+        auto devOutput = output_.packed_accessor<scalar_t, 4>();
+
+        int outputPlaneSize = devOutput.size(2) * devOutput.size(3);
+        dim3 gridSize(THCCeilDiv(outputPlaneSize, 256),
+            devOutput.size(1),
+            devOutput.size(0));
+        dim3 blockSize(outputPlaneSize > 256 ? 256 : outputPlaneSize);
```

Comment:
Hmm... in the original, it wasn't necessary to copy paste the block here and below. I guess the problem is that there is no default constructor in PackedTensorAccessor. Not sure if we should add one. I guess this is OK for now.

Reply:
I think I'm going to ask that you use a ternary here for the definition of `devInput` and `devOutput`, so you can avoid having to duplicate the epilogue. Something like: `auto devInput = (numInputDims == 3 ? input.unsqueeze(0) :  input).packed_accessor<scalar_t, 4>()` and likewise for the output